<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html lang="en">
<head>
<title>Source code</title>
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
</head>
<body>
<div class="sourceContainer">
<pre><span class="sourceLineNo">001</span>/**<a name="line.1"></a>
<span class="sourceLineNo">002</span> * Licensed to the Apache Software Foundation (ASF) under one<a name="line.2"></a>
<span class="sourceLineNo">003</span> * or more contributor license agreements.  See the NOTICE file<a name="line.3"></a>
<span class="sourceLineNo">004</span> * distributed with this work for additional information<a name="line.4"></a>
<span class="sourceLineNo">005</span> * regarding copyright ownership.  The ASF licenses this file<a name="line.5"></a>
<span class="sourceLineNo">006</span> * to you under the Apache License, Version 2.0 (the<a name="line.6"></a>
<span class="sourceLineNo">007</span> * "License"); you may not use this file except in compliance<a name="line.7"></a>
<span class="sourceLineNo">008</span> * with the License.  You may obtain a copy of the License at<a name="line.8"></a>
<span class="sourceLineNo">009</span> *<a name="line.9"></a>
<span class="sourceLineNo">010</span> *     http://www.apache.org/licenses/LICENSE-2.0<a name="line.10"></a>
<span class="sourceLineNo">011</span> *<a name="line.11"></a>
<span class="sourceLineNo">012</span> * Unless required by applicable law or agreed to in writing, software<a name="line.12"></a>
<span class="sourceLineNo">013</span> * distributed under the License is distributed on an "AS IS" BASIS,<a name="line.13"></a>
<span class="sourceLineNo">014</span> * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.<a name="line.14"></a>
<span class="sourceLineNo">015</span> * See the License for the specific language governing permissions and<a name="line.15"></a>
<span class="sourceLineNo">016</span> * limitations under the License.<a name="line.16"></a>
<span class="sourceLineNo">017</span> */<a name="line.17"></a>
<span class="sourceLineNo">018</span>package org.apache.hadoop.hbase.util;<a name="line.18"></a>
<span class="sourceLineNo">019</span><a name="line.19"></a>
<span class="sourceLineNo">020</span>import java.io.IOException;<a name="line.20"></a>
<span class="sourceLineNo">021</span>import java.io.InterruptedIOException;<a name="line.21"></a>
<span class="sourceLineNo">022</span>import java.util.concurrent.CompletableFuture;<a name="line.22"></a>
<span class="sourceLineNo">023</span>import java.util.concurrent.CompletionException;<a name="line.23"></a>
<span class="sourceLineNo">024</span>import java.util.concurrent.ExecutionException;<a name="line.24"></a>
<span class="sourceLineNo">025</span>import java.util.concurrent.Executor;<a name="line.25"></a>
<span class="sourceLineNo">026</span>import java.util.concurrent.Future;<a name="line.26"></a>
<span class="sourceLineNo">027</span>import java.util.concurrent.TimeUnit;<a name="line.27"></a>
<span class="sourceLineNo">028</span>import java.util.concurrent.TimeoutException;<a name="line.28"></a>
<span class="sourceLineNo">029</span>import java.util.function.BiConsumer;<a name="line.29"></a>
<span class="sourceLineNo">030</span>import org.apache.hadoop.hbase.exceptions.TimeoutIOException;<a name="line.30"></a>
<span class="sourceLineNo">031</span>import org.apache.yetus.audience.InterfaceAudience;<a name="line.31"></a>
<span class="sourceLineNo">032</span>import org.slf4j.Logger;<a name="line.32"></a>
<span class="sourceLineNo">033</span>import org.slf4j.LoggerFactory;<a name="line.33"></a>
<span class="sourceLineNo">034</span><a name="line.34"></a>
<span class="sourceLineNo">035</span>/**<a name="line.35"></a>
<span class="sourceLineNo">036</span> * Helper class for processing futures.<a name="line.36"></a>
<span class="sourceLineNo">037</span> */<a name="line.37"></a>
<span class="sourceLineNo">038</span>@InterfaceAudience.Private<a name="line.38"></a>
<span class="sourceLineNo">039</span>public final class FutureUtils {<a name="line.39"></a>
<span class="sourceLineNo">040</span><a name="line.40"></a>
<span class="sourceLineNo">041</span>  private static final Logger LOG = LoggerFactory.getLogger(FutureUtils.class);<a name="line.41"></a>
<span class="sourceLineNo">042</span><a name="line.42"></a>
<span class="sourceLineNo">043</span>  private FutureUtils() {<a name="line.43"></a>
<span class="sourceLineNo">044</span>  }<a name="line.44"></a>
<span class="sourceLineNo">045</span><a name="line.45"></a>
<span class="sourceLineNo">046</span>  /**<a name="line.46"></a>
<span class="sourceLineNo">047</span>   * This is method is used when you just want to add a listener to the given future. We will call<a name="line.47"></a>
<span class="sourceLineNo">048</span>   * {@link CompletableFuture#whenComplete(BiConsumer)} to register the {@code action} to the<a name="line.48"></a>
<span class="sourceLineNo">049</span>   * {@code future}. Ignoring the return value of a Future is considered as a bad practice as it may<a name="line.49"></a>
<span class="sourceLineNo">050</span>   * suppress exceptions thrown from the code that completes the future, and this method will catch<a name="line.50"></a>
<span class="sourceLineNo">051</span>   * all the exception thrown from the {@code action} to catch possible code bugs.<a name="line.51"></a>
<span class="sourceLineNo">052</span>   * &lt;p/&gt;<a name="line.52"></a>
<span class="sourceLineNo">053</span>   * And the error phone check will always report FutureReturnValueIgnored because every method in<a name="line.53"></a>
<span class="sourceLineNo">054</span>   * the {@link CompletableFuture} class will return a new {@link CompletableFuture}, so you always<a name="line.54"></a>
<span class="sourceLineNo">055</span>   * have one future that has not been checked. So we introduce this method and add a suppress<a name="line.55"></a>
<span class="sourceLineNo">056</span>   * warnings annotation here.<a name="line.56"></a>
<span class="sourceLineNo">057</span>   */<a name="line.57"></a>
<span class="sourceLineNo">058</span>  @SuppressWarnings("FutureReturnValueIgnored")<a name="line.58"></a>
<span class="sourceLineNo">059</span>  public static &lt;T&gt; void addListener(CompletableFuture&lt;T&gt; future,<a name="line.59"></a>
<span class="sourceLineNo">060</span>      BiConsumer&lt;? super T, ? super Throwable&gt; action) {<a name="line.60"></a>
<span class="sourceLineNo">061</span>    future.whenComplete((resp, error) -&gt; {<a name="line.61"></a>
<span class="sourceLineNo">062</span>      try {<a name="line.62"></a>
<span class="sourceLineNo">063</span>        // See this post on stack overflow(shorten since the url is too long),<a name="line.63"></a>
<span class="sourceLineNo">064</span>        // https://s.apache.org/completionexception<a name="line.64"></a>
<span class="sourceLineNo">065</span>        // For a chain of CompleableFuture, only the first child CompletableFuture can get the<a name="line.65"></a>
<span class="sourceLineNo">066</span>        // original exception, others will get a CompletionException, which wraps the original<a name="line.66"></a>
<span class="sourceLineNo">067</span>        // exception. So here we unwrap it before passing it to the callback action.<a name="line.67"></a>
<span class="sourceLineNo">068</span>        action.accept(resp, unwrapCompletionException(error));<a name="line.68"></a>
<span class="sourceLineNo">069</span>      } catch (Throwable t) {<a name="line.69"></a>
<span class="sourceLineNo">070</span>        LOG.error("Unexpected error caught when processing CompletableFuture", t);<a name="line.70"></a>
<span class="sourceLineNo">071</span>      }<a name="line.71"></a>
<span class="sourceLineNo">072</span>    });<a name="line.72"></a>
<span class="sourceLineNo">073</span>  }<a name="line.73"></a>
<span class="sourceLineNo">074</span><a name="line.74"></a>
<span class="sourceLineNo">075</span>  /**<a name="line.75"></a>
<span class="sourceLineNo">076</span>   * Almost the same with {@link #addListener(CompletableFuture, BiConsumer)} method above, the only<a name="line.76"></a>
<span class="sourceLineNo">077</span>   * exception is that we will call<a name="line.77"></a>
<span class="sourceLineNo">078</span>   * {@link CompletableFuture#whenCompleteAsync(BiConsumer, Executor)}.<a name="line.78"></a>
<span class="sourceLineNo">079</span>   * @see #addListener(CompletableFuture, BiConsumer)<a name="line.79"></a>
<span class="sourceLineNo">080</span>   */<a name="line.80"></a>
<span class="sourceLineNo">081</span>  @SuppressWarnings("FutureReturnValueIgnored")<a name="line.81"></a>
<span class="sourceLineNo">082</span>  public static &lt;T&gt; void addListener(CompletableFuture&lt;T&gt; future,<a name="line.82"></a>
<span class="sourceLineNo">083</span>      BiConsumer&lt;? super T, ? super Throwable&gt; action, Executor executor) {<a name="line.83"></a>
<span class="sourceLineNo">084</span>    future.whenCompleteAsync((resp, error) -&gt; {<a name="line.84"></a>
<span class="sourceLineNo">085</span>      try {<a name="line.85"></a>
<span class="sourceLineNo">086</span>        action.accept(resp, unwrapCompletionException(error));<a name="line.86"></a>
<span class="sourceLineNo">087</span>      } catch (Throwable t) {<a name="line.87"></a>
<span class="sourceLineNo">088</span>        LOG.error("Unexpected error caught when processing CompletableFuture", t);<a name="line.88"></a>
<span class="sourceLineNo">089</span>      }<a name="line.89"></a>
<span class="sourceLineNo">090</span>    }, executor);<a name="line.90"></a>
<span class="sourceLineNo">091</span>  }<a name="line.91"></a>
<span class="sourceLineNo">092</span><a name="line.92"></a>
<span class="sourceLineNo">093</span>  /**<a name="line.93"></a>
<span class="sourceLineNo">094</span>   * Return a {@link CompletableFuture} which is same with the given {@code future}, but execute all<a name="line.94"></a>
<span class="sourceLineNo">095</span>   * the callbacks in the given {@code executor}.<a name="line.95"></a>
<span class="sourceLineNo">096</span>   */<a name="line.96"></a>
<span class="sourceLineNo">097</span>  public static &lt;T&gt; CompletableFuture&lt;T&gt; wrapFuture(CompletableFuture&lt;T&gt; future,<a name="line.97"></a>
<span class="sourceLineNo">098</span>      Executor executor) {<a name="line.98"></a>
<span class="sourceLineNo">099</span>    CompletableFuture&lt;T&gt; wrappedFuture = new CompletableFuture&lt;&gt;();<a name="line.99"></a>
<span class="sourceLineNo">100</span>    addListener(future, (r, e) -&gt; {<a name="line.100"></a>
<span class="sourceLineNo">101</span>      if (e != null) {<a name="line.101"></a>
<span class="sourceLineNo">102</span>        wrappedFuture.completeExceptionally(e);<a name="line.102"></a>
<span class="sourceLineNo">103</span>      } else {<a name="line.103"></a>
<span class="sourceLineNo">104</span>        wrappedFuture.complete(r);<a name="line.104"></a>
<span class="sourceLineNo">105</span>      }<a name="line.105"></a>
<span class="sourceLineNo">106</span>    }, executor);<a name="line.106"></a>
<span class="sourceLineNo">107</span>    return wrappedFuture;<a name="line.107"></a>
<span class="sourceLineNo">108</span>  }<a name="line.108"></a>
<span class="sourceLineNo">109</span><a name="line.109"></a>
<span class="sourceLineNo">110</span>  /**<a name="line.110"></a>
<span class="sourceLineNo">111</span>   * Get the cause of the {@link Throwable} if it is a {@link CompletionException}.<a name="line.111"></a>
<span class="sourceLineNo">112</span>   */<a name="line.112"></a>
<span class="sourceLineNo">113</span>  public static Throwable unwrapCompletionException(Throwable error) {<a name="line.113"></a>
<span class="sourceLineNo">114</span>    if (error instanceof CompletionException) {<a name="line.114"></a>
<span class="sourceLineNo">115</span>      Throwable cause = error.getCause();<a name="line.115"></a>
<span class="sourceLineNo">116</span>      if (cause != null) {<a name="line.116"></a>
<span class="sourceLineNo">117</span>        return cause;<a name="line.117"></a>
<span class="sourceLineNo">118</span>      }<a name="line.118"></a>
<span class="sourceLineNo">119</span>    }<a name="line.119"></a>
<span class="sourceLineNo">120</span>    return error;<a name="line.120"></a>
<span class="sourceLineNo">121</span>  }<a name="line.121"></a>
<span class="sourceLineNo">122</span><a name="line.122"></a>
<span class="sourceLineNo">123</span>  // This method is used to record the stack trace that calling the FutureUtils.get method. As in<a name="line.123"></a>
<span class="sourceLineNo">124</span>  // async client, the retry will be done in the retry timer thread, so the exception we get from<a name="line.124"></a>
<span class="sourceLineNo">125</span>  // the CompletableFuture will have a stack trace starting from the root of the retry timer. If we<a name="line.125"></a>
<span class="sourceLineNo">126</span>  // just throw this exception out when calling future.get(by unwrapping the ExecutionException),<a name="line.126"></a>
<span class="sourceLineNo">127</span>  // the upper layer even can not know where is the exception thrown...<a name="line.127"></a>
<span class="sourceLineNo">128</span>  // See HBASE-22316.<a name="line.128"></a>
<span class="sourceLineNo">129</span>  private static void setStackTrace(Throwable error) {<a name="line.129"></a>
<span class="sourceLineNo">130</span>    StackTraceElement[] localStackTrace = Thread.currentThread().getStackTrace();<a name="line.130"></a>
<span class="sourceLineNo">131</span>    StackTraceElement[] originalStackTrace = error.getStackTrace();<a name="line.131"></a>
<span class="sourceLineNo">132</span>    StackTraceElement[] newStackTrace =<a name="line.132"></a>
<span class="sourceLineNo">133</span>      new StackTraceElement[localStackTrace.length + originalStackTrace.length + 1];<a name="line.133"></a>
<span class="sourceLineNo">134</span>    System.arraycopy(localStackTrace, 0, newStackTrace, 0, localStackTrace.length);<a name="line.134"></a>
<span class="sourceLineNo">135</span>    newStackTrace[localStackTrace.length] =<a name="line.135"></a>
<span class="sourceLineNo">136</span>      new StackTraceElement("--------Future", "get--------", null, -1);<a name="line.136"></a>
<span class="sourceLineNo">137</span>    System.arraycopy(originalStackTrace, 0, newStackTrace, localStackTrace.length + 1,<a name="line.137"></a>
<span class="sourceLineNo">138</span>      originalStackTrace.length);<a name="line.138"></a>
<span class="sourceLineNo">139</span>    error.setStackTrace(newStackTrace);<a name="line.139"></a>
<span class="sourceLineNo">140</span>  }<a name="line.140"></a>
<span class="sourceLineNo">141</span><a name="line.141"></a>
<span class="sourceLineNo">142</span>  /**<a name="line.142"></a>
<span class="sourceLineNo">143</span>   * If we could propagate the given {@code error} directly, we will fill the stack trace with the<a name="line.143"></a>
<span class="sourceLineNo">144</span>   * current thread's stack trace so it is easier to trace where is the exception thrown. If not, we<a name="line.144"></a>
<span class="sourceLineNo">145</span>   * will just create a new IOException and then throw it.<a name="line.145"></a>
<span class="sourceLineNo">146</span>   */<a name="line.146"></a>
<span class="sourceLineNo">147</span>  public static IOException rethrow(Throwable error) throws IOException {<a name="line.147"></a>
<span class="sourceLineNo">148</span>    if (error instanceof IOException) {<a name="line.148"></a>
<span class="sourceLineNo">149</span>      setStackTrace(error);<a name="line.149"></a>
<span class="sourceLineNo">150</span>      throw (IOException) error;<a name="line.150"></a>
<span class="sourceLineNo">151</span>    } else if (error instanceof RuntimeException) {<a name="line.151"></a>
<span class="sourceLineNo">152</span>      setStackTrace(error);<a name="line.152"></a>
<span class="sourceLineNo">153</span>      throw (RuntimeException) error;<a name="line.153"></a>
<span class="sourceLineNo">154</span>    } else if (error instanceof Error) {<a name="line.154"></a>
<span class="sourceLineNo">155</span>      setStackTrace(error);<a name="line.155"></a>
<span class="sourceLineNo">156</span>      throw (Error) error;<a name="line.156"></a>
<span class="sourceLineNo">157</span>    } else {<a name="line.157"></a>
<span class="sourceLineNo">158</span>      throw new IOException(error);<a name="line.158"></a>
<span class="sourceLineNo">159</span>    }<a name="line.159"></a>
<span class="sourceLineNo">160</span>  }<a name="line.160"></a>
<span class="sourceLineNo">161</span><a name="line.161"></a>
<span class="sourceLineNo">162</span>  /**<a name="line.162"></a>
<span class="sourceLineNo">163</span>   * A helper class for getting the result of a Future, and convert the error to an<a name="line.163"></a>
<span class="sourceLineNo">164</span>   * {@link IOException}.<a name="line.164"></a>
<span class="sourceLineNo">165</span>   */<a name="line.165"></a>
<span class="sourceLineNo">166</span>  public static &lt;T&gt; T get(Future&lt;T&gt; future) throws IOException {<a name="line.166"></a>
<span class="sourceLineNo">167</span>    try {<a name="line.167"></a>
<span class="sourceLineNo">168</span>      return future.get();<a name="line.168"></a>
<span class="sourceLineNo">169</span>    } catch (InterruptedException e) {<a name="line.169"></a>
<span class="sourceLineNo">170</span>      throw (IOException) new InterruptedIOException().initCause(e);<a name="line.170"></a>
<span class="sourceLineNo">171</span>    } catch (ExecutionException e) {<a name="line.171"></a>
<span class="sourceLineNo">172</span>      throw rethrow(e.getCause());<a name="line.172"></a>
<span class="sourceLineNo">173</span>    }<a name="line.173"></a>
<span class="sourceLineNo">174</span>  }<a name="line.174"></a>
<span class="sourceLineNo">175</span><a name="line.175"></a>
<span class="sourceLineNo">176</span>  /**<a name="line.176"></a>
<span class="sourceLineNo">177</span>   * A helper class for getting the result of a Future, and convert the error to an<a name="line.177"></a>
<span class="sourceLineNo">178</span>   * {@link IOException}.<a name="line.178"></a>
<span class="sourceLineNo">179</span>   */<a name="line.179"></a>
<span class="sourceLineNo">180</span>  public static &lt;T&gt; T get(Future&lt;T&gt; future, long timeout, TimeUnit unit) throws IOException {<a name="line.180"></a>
<span class="sourceLineNo">181</span>    try {<a name="line.181"></a>
<span class="sourceLineNo">182</span>      return future.get(timeout, unit);<a name="line.182"></a>
<span class="sourceLineNo">183</span>    } catch (InterruptedException e) {<a name="line.183"></a>
<span class="sourceLineNo">184</span>      throw (IOException) new InterruptedIOException().initCause(e);<a name="line.184"></a>
<span class="sourceLineNo">185</span>    } catch (ExecutionException e) {<a name="line.185"></a>
<span class="sourceLineNo">186</span>      throw rethrow(e.getCause());<a name="line.186"></a>
<span class="sourceLineNo">187</span>    } catch (TimeoutException e) {<a name="line.187"></a>
<span class="sourceLineNo">188</span>      throw new TimeoutIOException(e);<a name="line.188"></a>
<span class="sourceLineNo">189</span>    }<a name="line.189"></a>
<span class="sourceLineNo">190</span>  }<a name="line.190"></a>
<span class="sourceLineNo">191</span><a name="line.191"></a>
<span class="sourceLineNo">192</span>  /**<a name="line.192"></a>
<span class="sourceLineNo">193</span>   * Returns a CompletableFuture that is already completed exceptionally with the given exception.<a name="line.193"></a>
<span class="sourceLineNo">194</span>   */<a name="line.194"></a>
<span class="sourceLineNo">195</span>  public static &lt;T&gt; CompletableFuture&lt;T&gt; failedFuture(Throwable e) {<a name="line.195"></a>
<span class="sourceLineNo">196</span>    CompletableFuture&lt;T&gt; future = new CompletableFuture&lt;&gt;();<a name="line.196"></a>
<span class="sourceLineNo">197</span>    future.completeExceptionally(e);<a name="line.197"></a>
<span class="sourceLineNo">198</span>    return future;<a name="line.198"></a>
<span class="sourceLineNo">199</span>  }<a name="line.199"></a>
<span class="sourceLineNo">200</span>}<a name="line.200"></a>




























































</pre>
</div>
</body>
</html>
